File: backend-python/routes/matches.py
from typing import Annotated
from fastapi import APIRouter, Query, Depends, HTTPException, status, Body
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from validations import ListMatcheQuery, CreateMatchPayload, UpdateScorePayload
from db import get_db, Match, MatchStatus
from utils import get_match_status, sync_match_status
from ws import broadcast_match_created, broadcast_score_update

match_router = APIRouter(prefix="/matches")

MAX_LIMIT = 100

@match_router.get("/")
async def list_matches(
    query: Annotated[ListMatcheQuery, Query()],
    db: Annotated[AsyncSession, Depends(get_db)]
):
    limit = min(query.limit or 50, MAX_LIMIT)

    try:
        sql = select(Match).order_by(Match.created_at).limit(limit)
        result = await db.execute(sql)
        matches = result.scalars().all()
        return { "data": matches }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": "Failed to list matches", "details": str(e)},
        )

@match_router.post("/", status_code=status.HTTP_201_CREATED)
async def create_match(
    payload: Annotated[CreateMatchPayload, Body()],
    db: Annotated[AsyncSession, Depends(get_db)]
):
    try:
        calculated_status = get_match_status(payload.start_time, payload.end_time)
        new_match = Match(
            sport=payload.sport,
            home_team=payload.home_team,
            away_team=payload.away_team,
            start_time=payload.start_time.replace(tzinfo=None),
            end_time=payload.end_time.replace(tzinfo=None),
            home_score=payload.home_score,
            away_score=payload.away_score,
            status=calculated_status,
        )
        db.add(new_match)
        await db.commit()
        await db.refresh(new_match)

        await broadcast_match_created(new_match.to_dict())

        return {"data": new_match}
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": "Failed to create match", "details": str(e)},
        )

@match_router.patch("/{match_id}/score")
async def update_score(
    match_id: int,
    payload: Annotated[UpdateScorePayload, Body()],
    db: Annotated[AsyncSession, Depends(get_db)]
):
    try:
        stmt = select(Match).where(Match.id == match_id).limit(1)
        result = await db.execute(stmt)
        existing = result.scalar()
        if not existing:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Match not found",
            )
        async def status_updater(next_status):
            existing.status = next_status
            await db.commit()
        await sync_match_status(existing, status_updater)
        
        if existing.status != MatchStatus.LIVE:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Match is not live",
            )
        
        # Mutate existing object tracked by the session (automatically updates)
        existing.home_score = payload.home_score
        existing.away_score = payload.away_score
        await db.commit()
        await db.refresh(existing)

        await broadcast_score_update(match_id, payload.home_score, payload.away_score)
        return { "data": existing }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": "Failed to update score", "details": str(e)},
        )

File: backend/src/routes/matches.rs
use axum::{Json, Router, extract::{Path, Query, State}, http::StatusCode, response::IntoResponse, routing::{get, patch}};
use chrono::Utc;
use sea_orm::{ActiveModelTrait, ActiveValue::Set, EntityTrait, QueryOrder, QuerySelect};

use crate::{common::{constants::MAX_LIMIT, response::{ErrorResponse, GenericResponse}, state::SharedState}, entity::{matches, sea_orm_active_enums::MatchStatus}, utils::match_status::{get_match_status, sync_match_status}, validation::matches::{CreateMatchPayload, ListMatchQuery, UpdateScorePayload}};
use crate::entity::prelude::Matches;

pub fn create_router() -> Router<SharedState> {
    Router::new()
        .route("/", get(list_matches).post(create_match))
        .route("/{id}/score", patch(update_match_score))
}

#[axum::debug_handler]
async fn list_matches(
    Query(query): Query<ListMatchQuery>,
    State(state): State<SharedState>
) -> impl IntoResponse {
    let limit = query.limit.unwrap_or(50).min(MAX_LIMIT);

    let result = Matches::find()
        .order_by_desc(matches::Column::CreatedAt)
        .limit(limit)
        .all(&state.db).await;

    match result {
        Ok(matches) => (StatusCode::OK, Json(GenericResponse { data: matches })).into_response(),
        Err(_) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse {
                error: "Failed to list matches.".into(),
                details: None
            })
        ).into_response()
    } 
}

#[axum::debug_handler]
async fn create_match(
    State(state): State<SharedState>,
    Json(payload): Json<CreateMatchPayload>,
) -> impl IntoResponse {
    let status = get_match_status(payload.start_time, payload.end_time);
    let new_match = matches::ActiveModel {
        sport: Set(payload.sport),
        home_team: Set(payload.home_team),
        away_team: Set(payload.away_team),
        home_score: Set(payload.home_score.unwrap_or(0)),
        away_score: Set(payload.away_score.unwrap_or(0)),
        start_time: Set(payload.start_time.naive_utc()),
        end_time: Set(payload.end_time.naive_utc()),
        status: Set(status),
        created_at: Set(Utc::now().naive_utc()),
        ..Default::default()
    };

    match new_match.insert(&state.db).await {
        Ok(inserted) => {
            let json_msg = serde_json::to_value(inserted.clone()).unwrap();
            state.connection.broadcast_match_created(json_msg);

            (
                StatusCode::CREATED,
                Json(GenericResponse { data: inserted }),
            ).into_response()
        }
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse {
                error: "Failed to create match.".into(),
                details: Some(e.to_string()),
            }),
        ).into_response()
    }
}

#[axum::debug_handler]
async fn update_match_score(
    Path(match_id): Path<i32>,
    State(state): State<SharedState>,
    Json(payload): Json<UpdateScorePayload>,
) -> impl IntoResponse {
    // 1. Locate existing match
    let match_lookup = Matches::find_by_id(match_id).one(&state.db).await;

    let mut existing_match = match match_lookup {
        Ok(Some(m)) => m,
        Ok(None) => {
            return (
                StatusCode::NOT_FOUND,
                Json(ErrorResponse {
                    error: "Match not found".into(),
                    details: None,
                }),
            ).into_response()
        }
        Err(_) => {
            return (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(ErrorResponse {
                    error: "Failed to update score".into(),
                    details: None,
                }),
            ).into_response()
        }
    };

    // 2. Synchronize match status dynamically
    let mut status_changed = false;
    let mut fresh_status = existing_match.status.clone();

    sync_match_status(&existing_match, |next_status| {
        status_changed = true;
        fresh_status = next_status;
    });

    if status_changed {
        let mut active: matches::ActiveModel = existing_match.clone().into();
        active.status = Set(fresh_status.clone());
        if let Ok(updated_model) = active.update(&state.db).await {
            existing_match = updated_model;
        }
    }

    // 3. Confirm match is currently Live
    if fresh_status != MatchStatus::Live {
        return (
            StatusCode::CONFLICT,
            Json(ErrorResponse {
                error: "Match is not live".into(),
                details: None,
            }),
        ).into_response();
    }

    // 4. Update the scores
    let mut active: matches::ActiveModel = existing_match.into();
    active.home_score = Set(payload.home_score);
    active.away_score = Set(payload.away_score);

    match active.update(&state.db).await {
        Ok(updated) => {
            // Call broadcast logic from the WebSocket loop here:
            state.connection.broadcast_score_update(match_id, updated.home_score, updated.away_score).await;
            
            (StatusCode::OK, Json(GenericResponse { data: updated })).into_response()
        }
        Err(_) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(ErrorResponse {
                error: "Failed to update score".into(),
                details: None,
            }),
        ).into_response()
    }
}